Author: Nina Boiko

In [1]:
import pandas as pd
import numpy as np


In [2]:
!unzip stack-overflow-developer-survey-2025.zip

Archive:  stack-overflow-developer-survey-2025.zip
  inflating: 2025_Developer_Survey_Tool.pdf  
  inflating: survey_results_public.csv  
  inflating: survey_results_schema.csv  


In [ ]:
survey = pd.read_csv('survey_results_public.csv', low_memory=FALSE)
schema = pd.read_csv('survey_results_schema.csv')

In [4]:
pd.set_option('display.max_columns', None)

In [5]:
survey.shape[0]

49191

In [6]:
schema.head()

qid          qname                                           question  \
0  QID18  TechEndorse_1  What attracts you to a technology or causes yo...   
1  QID18  TechEndorse_2  What attracts you to a technology or causes yo...   
2  QID18  TechEndorse_3  What attracts you to a technology or causes yo...   
3  QID18  TechEndorse_4  What attracts you to a technology or causes yo...   
4  QID18  TechEndorse_5  What attracts you to a technology or causes yo...   

  type                                      sub  sq_id  
0   RO  AI integration or AI Agent capabilities    1.0  
1   RO                          Easy-to-use API    2.0  
2   RO                  Robust and complete API    3.0  
3   RO     Customizable and manageable codebase    4.0  
4   RO                   Reputation for quality    5.0

In [7]:
survey.head()

ResponseId                      MainBranch              Age  \
0           1  I am a developer by profession  25-34 years old   
1           2  I am a developer by profession  25-34 years old   
2           3  I am a developer by profession  35-44 years old   
3           4  I am a developer by profession  35-44 years old   
4           5  I am a developer by profession  35-44 years old   

                                           EdLevel  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1              Associate degree (A.A., A.S., etc.)   
2     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

                                          Employment  \
0                                           Employed   
1                                           Employed   
2  Independent contractor, freelancer, or self-em...   
3                                           Employed   
4  Independent contractor, freelancer, or self-em...   

                                    EmploymentAddl  WorkExp  \
0  Caring for dependents (children, elderly, etc.)      8.0   
1                                              NaN      2.0   
2                                None of the above     10.0   
3                                None of the above      4.0   
4  Caring for dependents (children, elderly, etc.)     21.0   

                                     LearnCodeChoose  \
0  Yes, I am not new to coding but am learning ne...   
1  Yes, I am not new to coding but am learning ne...   
2  Yes, I am not new to coding but am learning ne...   
3  Yes, I am not new to coding but am learning ne...   
4  No, I am not new to coding and did not learn n...   

                                           LearnCode  \
0  Online Courses or Certification (includes all ...   
1  Online Courses or Certification (includes all ...   
2  Online Courses or Certification (includes all ...   
3  Other online resources (e.g. standard search, ...   
4                                                NaN   

                                         LearnCodeAI  \
0  Yes, I learned how to use AI-enabled tools for...   
1  Yes, I learned how to use AI-enabled tools for...   
2  Yes, I learned how to use AI-enabled tools for...   
3  Yes, I learned how to use AI-enabled tools for...   
4  Yes, I learned how to use AI-enabled tools for...   

                                          AILearnHow  YearsCode  \
0                AI CodeGen tools or AI-enabled apps       14.0   
1                AI CodeGen tools or AI-enabled apps       10.0   
2  AI CodeGen tools or AI-enabled apps;Technical ...       12.0   
3  AI CodeGen tools or AI-enabled apps;Videos (no...        5.0   
4  Technical documentation (is generated for/by t...       22.0   

                DevType                   OrgSize                  ICorPM  \
0     Developer, mobile        20 to 99 employees          People manager   
1   Developer, back-end      500 to 999 employees  Individual contributor   
2  Developer, front-end                       NaN                     NaN   
3   Developer, back-end  10,000 or more employees  Individual contributor   
4   Engineering manager                       NaN                     NaN   

                                          RemoteWork  \
0                                             Remote   
1  Hybrid (some in-person, leans heavy to flexibi...   
2                                                NaN   
3                                             Remote   
4                                                NaN   

                                   PurchaseInfluence  TechEndorseIntro  \
0  Yes, I influenced the purchase of a substantia...              Work   
1                                                 No  Personal Project   
2                                                 No              Work   
3                                                 No  P

Було підраховано загальну кількість респондентів в опитуванні Stack Overflow Developer Survey 2025. У дослідженні взяли участь 49 191 респондентів.

In [8]:
schema['qname'].head()

0    TechEndorse_1
1    TechEndorse_2
2    TechEndorse_3
3    TechEndorse_4
4    TechEndorse_5
Name: qname, dtype: object

In [9]:
questions = schema['qname'].dropna().unique()

In [10]:
len(questions)

139

In [11]:
questions_in_survey = [q for q in questions if q in survey.columns]
len(questions_in_survey)

126

In [12]:
complete_mask = survey[questions_in_survey].notna().all(axis=1)

In [13]:
complete_count = complete_mask.sum()
complete_count

np.int64(0)

1.Взяла список всіх питань опитування
2.Перевірила, які з них реально є у відповідях
3.Для кожного респондента перевірила чи заповнив він всі ці питання
4 Порахувала таких людей виявилось 0

In [14]:
survey['WorkExp'].head()

0     8.0
1     2.0
2    10.0
3     4.0
4    21.0
Name: WorkExp, dtype: float64

In [15]:
survey['WorkExp'].isna().sum()


np.int64(6298)

In [16]:
survey['WorkExp'].describe()

count    42893.000000
mean        13.367403
std         10.800117
min          1.000000
25%          5.000000
50%         10.000000
75%         20.000000
max        100.000000
Name: WorkExp, dtype: float64

In [17]:
result = pd.Series({'mean': round(survey['WorkExp'].mean(), 2),
    'median': survey['WorkExp'].median(),
    'mode': survey['WorkExp'].mode().iloc[0]})

In [18]:
print(result)

mean      13.37
median    10.00
mode      10.00
dtype: float64


Я перевірила колонку з досвідом роботи, врахувала пропущені значення та порахувала середнє, медіану та моду досвіду респондентів. В опитуванні 42 893 респонденти вказали свій досвід роботи. Середній досвід роботи становить приблизно 13.4 року, медіанний — 10 років, а найчастіше зустрічається значення 10 років.
Це означає, що більшість респондентів мають близько 10 років професійного досвіду.

In [19]:
[c for c in survey.columns if 'Remote' in c or 'Work' in c]

['WorkExp',
 'RemoteWork',
 'ToolCountWork',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'DatabaseHaveWorkedWith',
 'DatabaseWantToWorkWith',
 'PlatformHaveWorkedWith',
 'PlatformWantToWorkWith',
 'WebframeHaveWorkedWith',
 'WebframeWantToWorkWith',
 'DevEnvsHaveWorkedWith',
 'DevEnvsWantToWorkWith',
 'SOTagsHaveWorkedWith',
 'SOTagsWantToWorkWith',
 'OfficeStackAsyncHaveWorkedWith',
 'OfficeStackAsyncWantToWorkWith',
 'CommPlatformHaveWorkedWith',
 'CommPlatformWantToWorkWith',
 'AIModelsHaveWorkedWith',
 'AIModelsWantToWorkWith']

In [20]:
survey['RemoteWork'].value_counts(dropna=False).head(20)

RemoteWork
NaN                                                                             15411
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [21]:
remote_count = (survey['RemoteWork'] == 'Remote').sum()

In [22]:
print(remote_count)

10931


В опитуванні помітна популярність віддаленої роботи: 10931 респондентів працює повністю remote. Також значна частина обирає гібридний формат, тоді як офісний формат менш поширений. Це свідчить про зростаючий попит на гнучкі умови роботи серед розробників.

In [23]:
[c for c in survey.columns if 'Language' in c]

['LanguageChoice',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LanguageAdmired',
 'LanguagesHaveEntry',
 'LanguagesWantEntry']

In [24]:
python_mask = survey['LanguageHaveWorkedWith'].str.contains('Python',na=False)

In [25]:
python_percent = python_mask.mean()*100

In [26]:
print(python_percent)

37.53938728629221


In [27]:
round(python_percent,1)

np.float64(37.5)

In [28]:
print(f"Відсоток респондентів, які програмують на Python:{round(python_percent,1)}%")

Відсоток респондентів, які програмують на Python:37.5%


Я перевірила колонку з мовами програмування, які використовували респонденти. Далі я визначила, хто з них працює з Python та порахувала частку таких відповідей від загальної кількості. В результаті вийшло, що 37.5% респондентів використовують Python.

In [31]:
survey.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

In [32]:
survey['LearnCode'].head()

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
Name: LearnCode, dtype: object

In [33]:
online_mask = survey['LearnCode'].str.contains('Online',na=False)

In [34]:
online_count = online_mask.sum()

In [35]:
print(online_count)

10973


Я обрала колонку LearnCode, яка описує способи навчання програмуванню. За допомогою фільтрації я визначила респондентів, які навчалися через онлайн-курси, та підрахувала їх кількість. В результаті 10973 респондентів зазначили, що використовували онлайн-курси для навчання програмуванню.

In [36]:
python_mask = survey['LanguageHaveWorkedWith'].str.contains('Python',na=False)
df_py = survey[python_mask].copy()

In [37]:
df_py= df_py.dropna(subset=['ConvertedCompYearly','Country'])

In [38]:
py_prog_by_country = (df_py.groupby('Country')['ConvertedCompYearly']
    .agg(mean_compensation='mean', median_compensation='median')
    .sort_values('median_compensation', ascending=False))

In [39]:
py_prog_by_country.head(10)

mean_compensation  median_compensation
Country                                                         
Oman                          390135.000000             390135.0
Andorra                       226103.500000             226103.5
United States of America      173298.590211             150000.0
Israel                        135828.365385             142594.0
Switzerland                   156456.600000             142592.0
Nomadic                       120131.571429             139218.0
Ireland                       120523.918919             116015.0
Luxembourg                    116014.714286             109054.0
Kyrgyzstan                    106008.500000             106008.5
Belize                        102121.000000             102121.0

Я відфільтрувала респондентів, які працювали з Python (через LanguageHaveWorkedWith) далі прибрала рядки без значень компенсації та країни (ConvertedCompYearly, Country). Потім згрупувала дані за країнами та для кожної країни порахувала середню та медіанну річну компенсацію. Результат відсортувала за медіаною та вивела топ-10 країн.

In [40]:
survey.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

In [45]:
top5 = (survey[['ResponseId','EdLevel','ConvertedCompYearly']]
        .dropna(subset=['ConvertedCompYearly','EdLevel'])
        .sort_values('ConvertedCompYearly', ascending=False))





In [48]:
top5 = (survey[['ResponseId','EdLevel','ConvertedCompYearly']]
        .dropna(subset=['ConvertedCompYearly','EdLevel'])
        .sort_values('ConvertedCompYearly', ascending=False)
        .head(5))

top5

ResponseId                                          EdLevel  \
34267       34268              Associate degree (A.A., A.S., etc.)   
28700       28701  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
43143       43144              Associate degree (A.A., A.S., etc.)   
35353       35354     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
45971       45972  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

       ConvertedCompYearly  
34267           50000000.0  
28700           33552715.0  
43143           18387548.0  
35353           15430267.0  
45971           13921760.0

Очистила дані від пропусків у зарплаті та рівні освіти.Відсортувала респондентів за річною компенсацією у спадному порядку.
Вибрала топ-5 респондентів з найбільшою зарплатою.